### ISSUE TITLE
Definition of "primitive" types

### ISSUE THREAD

**Issue by igarashitm** [atlasmap/atlasmap#323]
AtlasMap has a confusion on the definition of "primitive":
1. Java primitives: (boolean , byte , char , short , int , long , float, double)
2. `JavaField.isPrimitive() == true`: (Java primitives and boxed of those + **java.lang.String**)
3. AtlasMap primitives: (other than FieldType.COMPLEX, i.e. `JavaField.isPrimitive() == true` + java.lang.Object + date/time classes)

The boundary between 2. and 3. is ambiguous. IMHO `java.lang.String` should be Atlas primitive. See if there's any problem if `JavaField.isPrimitive()` returns `false` for java.lang.String. Also consider if we want to introduce `JavaField.isAtlasPrimitive()`.

Also `FieldType.DECIMAL` is in the `PRIMITIVE_FIELDTYPES` list ATM, but it obviously shouldn't be.

**Comment by mattrpav** [comment]
@igarashitm: This should already be defined. At one point there were *ModelUtil class that had the isPrimitive() methods and the ability to convert from the FieldType <-> the Java Class. (I don't see them in the code base atm).

One note to consider: the definition of a "Atlas Primitive" _should_ extend beyond just Java. The intent is for it to the Atlas FieldTypes to be comprehensive across _all_ formats, not just Java. For example. FieldType.INTEGER also maps to XML xs:int and xs:integer.

Perhaps a definitive table of what the Atlas FieldType primitives represent from a data-level would communicate their meaning more clearly, vs relying on the solely Java representation.

For example:
FieldType.INTEGER is defined as a 64-bit signed whole number

**Comment by igarashitm** [comment]
What do you mean? are you saying `JavaField.isPrimitive()` should return `true`? why?

FieldType.INTEGER was obviously mapped to 32-bit, i.e. Java `int`. That's why I introduced `FieldType.BIG_INTEGER`. There was no corresponding field type as opposed to we already had `FieldType.DECIMAL` which maps to Java BigDecimal.

FYI, in XML module, both of `xs:int` and `xs:integer` were actually mapped to FieldType.INTEGER, but I changed `xs:integer` to be mapped to FieldType.BIG_INTEGER.

Basically the original field type mapping was inconsistent and needed to clean up.

**Comment by igarashitm** [comment]
And `FieldType.LONG` is the one mapped to 64-bit

**Comment by mattrpav** [comment]
> What do you mean? are you saying JavaField.isPrimitive() should return true? why?

No, there was a utility method that accepted parameters. My short-hand left out the method argument.  It was something along the lines of:

  boolean AtlasFieldUtil.isPrimitive(FieldType fieldType) 

.. and a corresponding utility method specific for Java classes 

   boolean AtlasJavaModelUtil.isPrimitive(Class<?> clazz);

 > FieldType.INTEGER was obviously mapped to 32-bit, i.e. Java int. That's why I introduced 
 > FieldType.BIG_INTEGER. There was no corresponding field type as opposed to we already had 
 > FieldType.DECIMAL which maps to Java BigDecimal.

Correct, in my rush while updating the ticket I left out the signed bit division. FieldType.INTEGER should be defined as a signed 32-bit whole number (4 bytes), FieldType.LONG is signed 64-bit whole number (8 bytes), etc.

XML maps cleanly, since the spec declares bit length for number fields.

JSON.number should default map to something along the lines of a Double to prevent overflows in various JSON implementations that would be on either side of an AtlasMapping conversion. Additionally, AtlasMap Json module should support a configurable behavior for how to marshal/unmarshal json.number. This would allow users to specify exact bit length and precision on a per-mapping basis. JSON.number does not have a consensus interpretation of the specification and various implementations have adopted different bit lengths and precision acuity.

 > Basically the original field type mapping was inconsistent and needed to clean up.

Curiously, how was it inconsistent? The utility methods encapsulated all the field type <-> java class information.

**Comment by igarashitm** [comment]
> Curiously, how was it inconsistent? The utility methods encapsulated all the field type <-> java class information.
All those I fixed. Again, that utility method is still saying that the `java.lang.String` is primitive. And that's what I intend to address with this issue.

**Comment by mattrpav** [comment]
String should be an “Atlas Pimitive”.
